File: leetcode/src/reverseNodesInK.cpp
/*
 * given a SLL, reverse k nodes once and keep on doing it. if count of nodes to reverse less than k, remain them.
 *
 * e.g 1->2->3->4->5,
 * k = 2, return 2->1->4->3->5
 * k = 3, return 3->2->1->4->5
 *
 * for {1}, k = 2 return {1}
 *
 * */
#include "../include/preliminary.h"

class Solution{
public:
    ListNode* reverseKGroups(ListNode *head, int k){
        if(!head || k < 2)    return head;

        ListNode *pre = NULL, *h = head;
        head = NULL;
        while(h){
            ListNode *ln = h;

            int i = 1;
            for(; i < k && ln->next; ++i){
                ln = ln->next;
            }
            if(i < k)    break;  // remaining nodes is less than k

            ListNode *post = ln->next;
            ListNode *nh = reverse(h, post);

            if(!pre){
                head = nh;
            }else{
                pre->next = nh;
            }

            pre = h;  // after reverse, h becomes exclusive prev of next partial SLL
            h = post; // inclusive head of next partial SLL
        }

        return head ? head : h;  // head == NULL means whole length is less than k, nothing happened
    }

private:
    /*
     * @param ListNode *h: inclusive head of this partial SLL
     *        ListNode *post: exclusive end of this partial SLL
     * @return: new head of reversed SLL
     * */
    ListNode* reverse(ListNode *h, ListNode *post){
        if(!h)    return NULL;

        ListNode *l1 = h, *l2 = h->next;
        while(l2 != post){  // once loop clause to process l2
            ListNode *after = l2->next;
            l2->next = l1;

            l1 = l2;
            l2 = after;
        }
        h->next = post;

        return l1;
    }
};

/* unit test is in ../cpp_unittest/reverseNodesInK_unittest */

File: leetcode/java/reverseNodesInK.java
/*
 * given a SLL, reverse every k nodes at a time and return is modified head
 * if rest nodes less than k, remain them
 * e.g, 1->2->3->4->5
 * k=2, 2->1->4->3->5
 * k=3, 3->2->1->4->5
 * {1}, k=2, => {1}
 * */
import java.io.*;
import java.util.*;

class ListNode{
    int val;
    ListNode next;
    ListNode(int x){
        val = x;
        next = null;
    }
}

public class reverseNodesInK{
    private void reverse(ListNode pre, ListNode h, ListNode e, ListNode post){
        if(h==null)        return;
        ListNode l1=h, l2=h.next, after = l2.next;
        while(l2 != post){  //l1->l2->after => l1<-l2  after
            l2.next = l1;
            if(after == null)    break;
            l1 = l2;
            l2 = after;
            after = l2.next;
        }
        if(pre != null)    pre.next = e;
        h.next = post;
        return;
    }

    public ListNode reverseKGroups(ListNode head, int k){
        if(head==null || k<2)    return head;
        ListNode pre = null, h=head, post = null;
        head = null;
        while(h != null){
            ListNode e = h;
            int i=0;
            for(i=1;i<k && e.next != null;++i){
                e = e.next;
            }
            if(i<k)        break;
            post = e.next;
            reverse(pre, h, e, post);
            if(head==null)    head = e;

            pre = h;    //move forward. after reverse, h becomes prev of post
            h=post;
        }
        if(head == null)    return h;
        return head;
    }

    public static void main(String[] args){
        return;
    }
}
